File: examples/holeEtching/holeEtching.cpp
#include <geometries/psMakeHole.hpp>
#include <models/psSF6O2Etching.hpp>

#include <process/psProcess.hpp>
#include <psUtil.hpp>

using namespace viennaps;

int main(int argc, char *argv[]) {
  using NumericType = double;
  constexpr int D = 3;

  omp_set_num_threads(16);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    // Try default config file
    params.readConfigFile("config.txt");
    if (params.m.empty()) {
      std::cout << "No configuration file provided!" << std::endl;
      std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
      return 1;
    }
  }

  // set parameter units
  units::Length::setUnit(params.get<std::string>("lengthUnit"));
  units::Time::setUnit(params.get<std::string>("timeUnit"));

  // geometry setup
  auto geometry = Domain<NumericType, D>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));
  MakeHole<NumericType, D>(geometry, params.get("holeRadius"),
                           0.0, // holeDepth
                           0.0, // holeTaperAngle
                           params.get("maskHeight"), params.get("taperAngle"),
                           HoleShape::HALF)
      .apply();

  // use pre-defined model SF6O2 etching model
  auto modelParams = SF6O2Etching<NumericType, D>::defaultParameters();
  modelParams.ionFlux = params.get("ionFlux");
  modelParams.etchantFlux = params.get("etchantFlux");
  modelParams.passivationFlux = params.get("oxygenFlux");
  modelParams.Ions.meanEnergy = params.get("meanEnergy");
  modelParams.Ions.sigmaEnergy = params.get("sigmaEnergy");
  modelParams.Ions.exponent = params.get("ionExponent");
  modelParams.Passivation.A_ie = params.get("A_O");
  modelParams.Substrate.A_ie = params.get("A_Si");
  modelParams.etchStopDepth = params.get("etchStopDepth");
  auto model = SmartPointer<SF6O2Etching<NumericType, D>>::New(modelParams);

  CoverageParameters coverageParams;
  coverageParams.tolerance = 1e-4;

  RayTracingParameters rayTracingParams;
  rayTracingParams.raysPerPoint = params.get<unsigned>("raysPerPoint");

  AdvectionParameters advectionParams;
  advectionParams.spatialScheme =
      util::convertSpatialScheme(params.get<std::string>("spatialScheme"));
  advectionParams.temporalScheme =
      util::convertTemporalScheme(params.get<std::string>("temporalScheme"));

  // process setup
  Process<NumericType, D> process(geometry, model);
  process.setProcessDuration(params.get("processTime"));
  process.setParameters(coverageParams);
  process.setParameters(rayTracingParams);
  process.setParameters(advectionParams);

  // print initial surface
  geometry->saveSurfaceMesh("initial.vtp");

  // run the process
  process.apply();

  // print final surface
  geometry->saveSurfaceMesh(params.get<std::string>("outputFile"), true, 0.01,
                            true);
}

File: examples/holeEtching/holeEtching.py
from argparse import ArgumentParser
import viennaps as ps

# parse config file name and simulation dimension
parser = ArgumentParser(prog="holeEtching", description="Run a hole etching process.")
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
args = parser.parse_args()

# switch between 2D and 3D mode
if args.dim == 2:
    print("Running 2D simulation.")
else:
    print("Running 3D simulation.")
ps.setDimension(args.dim)
ps.setNumThreads(16)

params = ps.readConfigFile(args.filename)

ps.Length.setUnit(params["lengthUnit"])
ps.Time.setUnit(params["timeUnit"])

def run_simulation(intermediate_velocities, suffix):
    # geometry setup, all units in um
    geometry = ps.Domain(
        gridDelta=params["gridDelta"],
        xExtent=params["xExtent"],
        yExtent=params["yExtent"],
    )
    
    hole_shape = ps.HoleShape.QUARTER if args.dim == 3 else ps.HoleShape.HALF

    ps.MakeHole(
        domain=geometry,
        holeRadius=params["holeRadius"],
        holeDepth=0.0,
        maskHeight=params["maskHeight"],
        maskTaperAngle=params["taperAngle"],
        holeShape=hole_shape,
    ).apply()

    # use pre-defined model SF6O2 etching model
    modelParams = ps.SF6O2Etching.defaultParameters()
    modelParams.ionFlux = params["ionFlux"]
    modelParams.etchantFlux = params["etchantFlux"]
    modelParams.passivationFlux = params["oxygenFlux"]
    modelParams.Ions.meanEnergy = params["meanEnergy"]
    modelParams.Ions.sigmaEnergy = params["sigmaEnergy"]
    modelParams.Ions.exponent = params["ionExponent"]
    modelParams.Passivation.A_ie = params["A_O"]
    modelParams.Substrate.A_ie = params["A_Si"]
    modelParams.etchStopDepth = params["etchStopDepth"]
    model = ps.SF6O2Etching(modelParams)

    covParams = ps.CoverageParameters()
    covParams.tolerance = 1e-4

    rayParams = ps.RayTracingParameters()
    rayParams.raysPerPoint = int(params["raysPerPoint"])

    advParams = ps.AdvectionParameters()
    advParams.spatialScheme = ps.util.convertSpatialScheme(
        params["spatialScheme"]
    )
    advParams.temporalScheme = ps.util.convertTemporalScheme(
        params["temporalScheme"]
    )
    advParams.calculateIntermediateVelocities = intermediate_velocities

    # process setup
    process = ps.Process(geometry, model)
    process.setProcessDuration(params["processTime"])  # seconds
    process.setParameters(covParams)
    process.setParameters(rayParams)
    process.setParameters(advParams)

    # print initial surface
    if suffix == "_noIntermediate":
        geometry.saveSurfaceMesh(filename="initial.vtp")

    # run the process
    process.apply()

    # print final surface
    output_file = params["outputFile"]
    if "." in output_file:
        parts = output_file.rsplit(".", 1)
        output_file = f"{parts[0]}{suffix}.{parts[1]}"
    else:
        output_file += suffix

    geometry.saveSurfaceMesh(filename=output_file, addInterfaces=True, boolMaterials=True)

print("Running simulation without intermediate velocity calculation...")
run_simulation(False, "_noIntermediate")

print("Running simulation with intermediate velocity calculation...")
run_simulation(True, "_intermediate")
